# Toronto Neighborhoods

First we need to install any packages that hasn't been previously installed. Once installed, I comment the lines.

In [1]:
# import sys
# !{sys.executable} -m pip install bs4
# !{sys.executable} -m pip install lxml

Then import the required libraries to parse the html page.

In [2]:
from bs4 import BeautifulSoup
from lxml import etree
import requests

All webparsing activities will go into the below cell.

In [3]:
quote_page = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
source = requests.get(quote_page).text
soup = BeautifulSoup(source, "lxml")
table = soup.find('table',{'class':'wikitable sortable'})

Now, the data from the html should be put into lists.

In [4]:
PostalCode = []
Borough = []
Neighborhood = []
for tr in table.find_all('tr'):
        tds = tr.find_all('td')
        if not tds:
            continue
        PC, B, N = [td.text.strip() for td in tds[:3]]
        PostalCode.append(PC)
        Borough.append(B)
        Neighborhood.append(N)

Pandas needs to be imported so that we can put the table data into a dataframe.

In [5]:
import pandas as pd
df = pd.DataFrame({'PostalCode': PostalCode,
                  'Borough': Borough,
                  'Neighborhood': Neighborhood})

The dataframe should match the requirements of the submission, which are explained inline below.

In [6]:
df = df[df.Borough != 'Not assigned'] # To ignore cells with a borough that are 'Not assigned'
nameless = df.Neighborhood == 'Not assigned' # To identify rows that have a Borough but do not have a neighborhood assigned
df.loc[nameless, 'Neighborhood'] = df.loc[nameless, 'Borough'] # To replace 'Not assigned' neighborhoods with the borough name
neighborhoods = df.groupby(['PostalCode','Borough'])['Neighborhood'].apply(', '.join).reset_index() # To group the PostalCode and Boroughs together and concatenate the neighborhoods
neighborhoods

PostalCode           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
..         ...               ...   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A      Queen's Park   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A         Etobicoke   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                          Neighborhood  
0                                       Rouge, Malvern  
1               Highland Creek, Rouge Hill, Port Union  
2                    Guildwood, Morningside, West Hill  
3                                               Woburn  
4                                            Cedarbrae  
5                                  Scarborough Village  
6          East Birchmount Park, Ionview, Kennedy Park  
7                      Clairlea, Golden Mile, Oakridge  
8      Cliffcrest, Cliffside, Scarborough Village West  
9                          Birch Cliff, Cliffside West  
10   Dorset Park, Scarborough Town Centre, Wexford ...  
11                                   Maryvale, Wexford  
12                                           Agincourt  
13             Clarks Corners, Sullivan, Tam O'Shanter  
14   Agincourt North, L'Amoreaux East, Milliken, St...  
15                                     L'Amoreaux West  
16                                         Upper Rouge  
17                                   Hillcrest Village  
18                        Fairview, Henry Farm, Oriole  
19                                     Bayview Village  
20                            Silver Hills, York Mills  
21                             Newtonbrook, Willowdale  
22                                    Willowdale South  
23                                     York Mills West  
24                                     Willowdale West  
25                                           Parkwoods  
26                                     Don Mills North  
27                    Flemingdon Park, Don Mills South  
28     Bathurst Manor, Downsview North, Wilson Heights  
29                     Northwood Park, York University  
..  

Finally, the below code retrieves the shape of our dataframe, "neighborhoods".

In [7]:
neighborhoods.shape

(103, 3)

We now can map latitude and longitude coordinates to each of the 103 PostalCodes.

In [8]:
# !{sys.executable} -m pip install geocoder
# decided not to use geocoder because it was taking too long to set up and work

In [9]:
coordinates = pd.read_csv("https://cocl.us/Geospatial_data")

In [10]:
neighborhoods['Latitude'] = coordinates['Latitude']
neighborhoods['Longitude'] = coordinates['Longitude']
neighborhoods

PostalCode           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
..         ...               ...   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A      Queen's Park   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A         Etobicoke   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                       Rouge, Malvern  43.806686 -79.194353  
1               Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497  
2                    Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
5                                  Scarborough Village  43.744734 -79.239476  
6          East Birchmount Park, Ionview, Kennedy Park  43.727929 -79.262029  
7                      Clairlea, Golden Mile, Oakridge  43.711112 -79.284577  
8      Cliffcrest, Cliffside, Scarborough Village West  43.716316 -79.239476  
9                          Birch Cliff, Cliffside West  43.692657 -79.264848  
10   Dorset Park, Scarborough Town Centre, Wexford ...  43.757410 -79.273304  
11                                   Maryvale, Wexford  43.750072 -79.295849  
12                                           Agincourt  43.794200 -79.262029  
13             Clarks Corners, Sullivan, Tam O'Shanter  43.781638 -79.304302  
14   Agincourt North, L'Amoreaux East, Milliken, St...  43.815252 -79.284577  
15                                     L'Amoreaux West  43.799525 -79.318389  
16                                         Upper Rouge  43.836125 -79.205636  
17                                   Hillcrest Village  43.803762 -79.363452  
18                        Fairview, Henry Farm, Oriole  43.778517 -79.346556  
19                                     Bayview Village  43.786947 -79.385975  
20                            Silver Hills, York Mills  43.757490 -79.374714  
21                             Ne

# Third Question
Here I will cluster the neighborhoods together. I'll focus on Etobicoke, because it sounds interesting and I can't pronounce it. First, the relevant libraries should be imported for clustering and visualization.

In [11]:
#!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

# !conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

Now we'll do some preparation for seeing the neighborhoods

In [12]:
etobicoke_data = neighborhoods[neighborhoods['Borough'] == 'Etobicoke'].reset_index(drop=True)
etobicoke_data.head()

PostalCode    Borough                                       Neighborhood  \
0        M8V  Etobicoke       Humber Bay Shores, Mimico South, New Toronto   
1        M8W  Etobicoke                             Alderwood, Long Branch   
2        M8X  Etobicoke      The Kingsway, Montgomery Road, Old Mill North   
3        M8Y  Etobicoke  Humber Bay, King's Mill Park, Kingsway Park So...   
4        M8Z  Etobicoke  Kingsway Park South West, Mimico NW, The Queen...   

    Latitude  Longitude  
0  43.605647 -79.501321  
1  43.602414 -79.543484  
2  43.653654 -79.506944  
3  43.636258 -79.498509  
4  43.628841 -79.520999

In [13]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="ca_explorer")
TR_location = geolocator.geocode(address)
TR_latitude = TR_location.latitude
TR_longitude = TR_location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(TR_latitude, TR_longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


In [14]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[TR_latitude, TR_longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(etobicoke_data['Latitude'], etobicoke_data['Longitude'], etobicoke_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Now we can pull some information from Foursquare to begin clustering the neighborhoods.

In [15]:
CLIENT_ID = '4GSJSOHDZKYYB0XYCN0K30QT1SNPDDNGE3YY0EZJOJAGOKYF' # your Foursquare ID
CLIENT_SECRET = '4V0KU1O3BADD0UX0ZZ4FFCH5BPSFRPIUACWN3CKJRSSBNLSD' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 4GSJSOHDZKYYB0XYCN0K30QT1SNPDDNGE3YY0EZJOJAGOKYF
CLIENT_SECRET:4V0KU1O3BADD0UX0ZZ4FFCH5BPSFRPIUACWN3CKJRSSBNLSD


In [16]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [17]:
LIMIT = 100
radius = 500

etobicoke_venues = getNearbyVenues(names=etobicoke_data['Neighborhood'],
                                   latitudes=etobicoke_data['Latitude'],
                                   longitudes=etobicoke_data['Longitude']
                                  )

Humber Bay Shores, Mimico South, New Toronto
Alderwood, Long Branch
The Kingsway, Montgomery Road, Old Mill North
Humber Bay, King's Mill Park, Kingsway Park South East, Mimico NE, Old Mill South, The Queensway East, Royal York South East, Sunnylea
Kingsway Park South West, Mimico NW, The Queensway West, Royal York South West, South of Bloor
Islington Avenue
Cloverdale, Islington, Martin Grove, Princess Gardens, West Deane Park
Bloordale Gardens, Eringate, Markland Wood, Old Burnhamthorpe
Westmount
Kingsview Village, Martin Grove Gardens, Richview Gardens, St. Phillips
Albion Gardens, Beaumond Heights, Humbergate, Jamestown, Mount Olive, Silverstone, South Steeles, Thistletown
Northwest


In [18]:
print(etobicoke_venues.shape)
etobicoke_venues.head()

(71, 7)


Neighborhood  Neighborhood Latitude  \
0  Humber Bay Shores, Mimico South, New Toronto              43.605647   
1  Humber Bay Shores, Mimico South, New Toronto              43.605647   
2  Humber Bay Shores, Mimico South, New Toronto              43.605647   
3  Humber Bay Shores, Mimico South, New Toronto              43.605647   
4  Humber Bay Shores, Mimico South, New Toronto              43.605647   

   Neighborhood Longitude                     Venue  Venue Latitude  \
0              -79.501321                      LCBO       43.602281   
1              -79.501321            Domino's Pizza       43.601676   
2              -79.501321  New Toronto Fish & Chips       43.601849   
3              -79.501321   Delicia Bakery & Pastry       43.601403   
4              -79.501321     Lucky Dice Restaurant       43.601392   

   Venue Longitude Venue Category  
0       -79.499302   Liquor Store  
1       -79.500908    Pizza Place  
2       -79.503281     Restaurant  
3       -79.503012         Bakery  
4       -79.503056           Café

In [19]:
etobicoke_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Albion Gardens, Beaumond Heights, Humbergate, J...                     11   
Alderwood, Long Branch                                                 10   
Bloordale Gardens, Eringate, Markland Wood, Old...                      8   
Cloverdale, Islington, Martin Grove, Princess G...                      1   
Humber Bay Shores, Mimico South, New Toronto                           13   
Humber Bay, King's Mill Park, Kingsway Park Sou...                      1   
Kingsview Village, Martin Grove Gardens, Richvi...                      4   
Kingsway Park South West, Mimico NW, The Queens...                     11   
Northwest                                                               3   
The Kingsway, Montgomery Road, Old Mill North                           3   
Westmount                                                               6   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Albion Gardens, Beaumond Heights, Humbergate, J...                      11   
Alderwood, Long Branch                                                  10   
Bloordale Gardens, Eringate, Markland Wood, Old...                       8   
Cloverdale, Islington, Martin Grove, Princess G...                       1   
Humber Bay Shores, Mimico South, New Toronto                            13   
Humber Bay, King's Mill Park, Kingsway Park Sou...                       1   
Kingsview Village, Martin Grove Gardens, Richvi...                       4   
Kingsway Park South West, Mimico NW, The Queens...                      11   
Northwest                                                                3   
The Kingsway, Montgomery Road, Old Mill North                            3   
Westmount                                                                6   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Albion Gardens, Beaumond Heights, Humbergate, J...     11              11   
Alderwood, Long Branch                                 10              10   
Bloordale Gardens, Eringate, Markland Wood, Old...      8               8   
Cloverdale, Islington, Martin Grove, Princess G...      1               1   
Humber Bay Shores, Mimico South, New Toronto           13              13   
Humber Bay, King's Mill Park, Kingsway Park Sou...      1               1   
Kingsview Village, Martin Grove Gardens, Richvi...      4               4   
Kingsway Park South West, Mimico NW, The Queens...     11              11   
Northwest                                               3               3   
The Kingsway, Montgomery Road, Old Mill North           3               3   
Westmount                                               6               6   

                                                    Venue Longitude  \
Neighborhood                                                          
Albion Gardens, Beaumond Heights, Humbergate, J...               11   
Alderwood, Long Branch                                           10   
Bloordale Gardens, Eringate, Markland Wood, Old...                8   
Cloverdale, Islington, Martin Grove, Princess G...                1   
Humber Bay Shores, Mimico South, New Toronto                     13   
Humber Bay, King's Mill Park, Kingsway Park Sou...                1   
Kingsview Village, Martin Grove Gardens, Richvi...                4   
Kingsway Park South West, Mimico NW, The Queens...               11   
Northwest                                                         3   
The Kingsway, Montgomery Road, Old Mill North                     3   
Westmount                                                         6   

                                                    Venue Category  
Neighborhood                            

In [20]:
print('There are {} uniques categories.'.format(len(etobicoke_venues['Venue Category'].unique())))

There are 40 uniques categories.


In [21]:
# one hot encoding
etobicoke_onehot = pd.get_dummies(etobicoke_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
etobicoke_onehot['Neighborhood'] = etobicoke_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [etobicoke_onehot.columns[-1]] + list(etobicoke_onehot.columns[:-1])
etobicoke_onehot = etobicoke_onehot[fixed_columns]

etobicoke_onehot.head()

Neighborhood  American Restaurant  \
0  Humber Bay Shores, Mimico South, New Toronto                    0   
1  Humber Bay Shores, Mimico South, New Toronto                    0   
2  Humber Bay Shores, Mimico South, New Toronto                    0   
3  Humber Bay Shores, Mimico South, New Toronto                    0   
4  Humber Bay Shores, Mimico South, New Toronto                    0   

   Athletics & Sports  Bakery  Bank  Bar  Baseball Field  Beer Store  \
0                   0       0     0    0               0           0   
1                   0       0     0    0               0           0   
2                   0       0     0    0               0           0   
3                   0       1     0    0               0           0   
4                   0       0     0    0               0           0   

   Burger Joint  Bus Line     ...       Rental Car Location  Restaurant  \
0             0         0     ...                         0           0   
1             0         0     ...                         0           0   
2             0         0     ...                         0           1   
3             0         0     ...                         0           0   
4             0         0     ...                         0           0   

   River  Sandwich Place  Shopping Plaza  Skating Rink  Supplement Shop  \
0      0               0               0             0                0   
1      0               0               0             0                0   
2      0               0               0             0                0   
3      0               0               0             0                0   
4      0               0               0             0                0   

   Thrift / Vintage Store  Video Store  Wings Joint  
0                       0            0            0  
1                       0            0            0  
2                       0            0            0  
3                       0            0            0  
4                       0            0            0  

[5 rows x 41 columns]

In [22]:
etobicoke_grouped = etobicoke_onehot.groupby('Neighborhood').mean().reset_index()
etobicoke_grouped

Neighborhood  American Restaurant  \
0   Albion Gardens, Beaumond Heights, Humbergate, ...             0.000000   
1                              Alderwood, Long Branch             0.000000   
2   Bloordale Gardens, Eringate, Markland Wood, Ol...             0.000000   
3   Cloverdale, Islington, Martin Grove, Princess ...             0.000000   
4        Humber Bay Shores, Mimico South, New Toronto             0.076923   
5   Humber Bay, King's Mill Park, Kingsway Park So...             0.000000   
6   Kingsview Village, Martin Grove Gardens, Richv...             0.000000   
7   Kingsway Park South West, Mimico NW, The Queen...             0.000000   
8                                           Northwest             0.000000   
9       The Kingsway, Montgomery Road, Old Mill North             0.000000   
10                                          Westmount             0.000000   

    Athletics & Sports    Bakery  Bank       Bar  Baseball Field  Beer Store  \
0                  0.0  0.000000   0.0  0.000000             0.0    0.090909   
1                  0.1  0.000000   0.0  0.000000             0.0    0.000000   
2                  0.0  0.000000   0.0  0.000000             0.0    0.125000   
3                  0.0  0.000000   1.0  0.000000             0.0    0.000000   
4                  0.0  0.076923   0.0  0.000000             0.0    0.000000   
5                  0.0  0.000000   0.0  0.000000             1.0    0.000000   
6                  0.0  0.000000   0.0  0.000000             0.0    0.000000   
7                  0.0  0.090909   0.0  0.000000             0.0    0.000000   
8                  0.0  0.000000   0.0  0.333333             0.0    0.000000   
9                  0.0  0.000000   0.0  0.000000             0.0    0.000000   
10                 0.0  0.000000   0.0  0.000000             0.0    0.000000   

    Burger Joint  Bus Line     ...       Rental Car Location  Restaurant  \
0       0.000000      0.00     ...                  0.000000    0.000000   
1       0.000000      0.00     ...                  0.000000    0.000000   
2       0.000000      0.00     ...                  0.000000    0.000000   
3       0.000000      0.00     ...                  0.000000    0.000000   
4       0.000000      0.00     ...                  0.000000    0.076923   
5       0.000000      0.00     ...                  0.000000    0.000000   
6       0.000000      0.25     ...                  0.000000    0.000000   
7       0.090909      0.00     ...                  0.000000    0.000000   
8       0.000000      0.00     ...                  0.333333    0.000000   
9       0.000000      0.00     ...                  0.000000    0.000000   
10      0.000000      0.00     ...                  0.000000    0.000000   

       River  Sandwich Place  Shopping Plaza  Skating Rink  Supplement Shop  \
0   0.000000        0.090909           0.000           0.0         0.000000   
1   0.000000        0.100000           0.000           0.1         0.000000   
2   0.000000        0.000000           0.125           0.0         0.000000   
3   0.000000        0.000000           0.000           0.0         0.000000   
4   0.000000        0.076923           0.000           0.0         0.000000   
5   0.000000        0.000000           0.000           0.0         0.000000   
6   0.000000        0.000000           0.000           0.0         0.000000   
7   0.000000        0.090909           0.000           0.0         0.090909   
8   0.000000        0.000000           0.000           0.0         0.000000   
9   0.333333        0.000000           0.000           0.0         0.000000   
10  0.000000        0.166667           0.000           0.0         0.000000   

    Thrift / Vintage Store  Video Store  Wings Joint  
0                 0.000000     0.090909     0.000000  
1                 0.000000     0.000000     0.000000  
2                 0.000000     0.000000     0.000000  
3                 0.000000     0.000000     0.000000  
4     

In [23]:
num_top_venues = 5

for hood in etobicoke_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = etobicoke_grouped[etobicoke_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Albion Gardens, Beaumond Heights, Humbergate, Jamestown, Mount Olive, Silverstone, South Steeles, Thistletown----
                 venue  freq
0        Grocery Store  0.18
1          Coffee Shop  0.09
2       Sandwich Place  0.09
3             Pharmacy  0.09
4  Fried Chicken Joint  0.09


----Alderwood, Long Branch----
                venue  freq
0         Pizza Place   0.2
1  Athletics & Sports   0.1
2                 Pub   0.1
3                 Gym   0.1
4        Skating Rink   0.1


----Bloordale Gardens, Eringate, Markland Wood, Old Burnhamthorpe----
            venue  freq
0            Park  0.12
1        Pharmacy  0.12
2     Pizza Place  0.12
3      Beer Store  0.12
4  Shopping Plaza  0.12


----Cloverdale, Islington, Martin Grove, Princess Gardens, West Deane Park----
                       venue  freq
0                       Bank   1.0
1        American Restaurant   0.0
2        Rental Car Location   0.0
3  Middle Eastern Restaurant   0.0
4          Mobile Phone Shop   0.0


## Well that's interesting
In the PostalCode associated with Humber Bay etc., it has only a baseball field. That's it. Similarly, that associated with Cloverdale only has a bank. That will be interesting to see later, and I can imagine it will receive its own cluster later. In either case, it either suggests a lack of Foursquare use in Etobicoke or more simply that there is just not much around these areas.

Press on!

In [24]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [25]:
import numpy as np

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = etobicoke_grouped['Neighborhood']

for ind in np.arange(etobicoke_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(etobicoke_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0  Albion Gardens, Beaumond Heights, Humbergate, ...         Grocery Store   
1                             Alderwood, Long Branch           Pizza Place   
2  Bloordale Gardens, Eringate, Markland Wood, Ol...                  Café   
3  Cloverdale, Islington, Martin Grove, Princess ...                  Bank   
4       Humber Bay Shores, Mimico South, New Toronto                   Gym   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0   Fried Chicken Joint           Video Store        Sandwich Place   
1                   Gym        Sandwich Place           Coffee Shop   
2           Pizza Place     Convenience Store        Shopping Plaza   
3           Wings Joint    Chinese Restaurant   Fried Chicken Joint   
4           Pizza Place                Bakery                  Café   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0          Liquor Store            Beer Store           Coffee Shop   
1              Pharmacy                   Pub                  Pool   
2          Liquor Store            Beer Store                  Park   
3           Flower Shop  Fast Food Restaurant             Drugstore   
4  Fast Food Restaurant           Flower Shop   Fried Chicken Joint   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0  Fast Food Restaurant           Pizza Place               Pharmacy  
1          Skating Rink    Athletics & Sports                    Bar  
2              Pharmacy             Drugstore         Discount Store  
3        Discount Store     Convenience Store            Coffee Shop  
4            Hobby Shop          Liquor Store               Pharmacy

In [26]:
# set number of clusters
kclusters = 5

etobicoke_grouped_clustering = etobicoke_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(etobicoke_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([3, 3, 3, 2, 3, 0, 3, 3, 4, 1], dtype=int32)

In [27]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

etobicoke_merged = etobicoke_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
etobicoke_merged = etobicoke_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
etobicoke_merged = etobicoke_merged[np.isfinite(etobicoke_merged['Cluster Labels'])]

etobicoke_merged['Cluster Labels'] = etobicoke_merged['Cluster Labels'].astype('int64')

etobicoke_merged # check the last columns!

PostalCode    Borough                                       Neighborhood  \
0         M8V  Etobicoke       Humber Bay Shores, Mimico South, New Toronto   
1         M8W  Etobicoke                             Alderwood, Long Branch   
2         M8X  Etobicoke      The Kingsway, Montgomery Road, Old Mill North   
3         M8Y  Etobicoke  Humber Bay, King's Mill Park, Kingsway Park So...   
4         M8Z  Etobicoke  Kingsway Park South West, Mimico NW, The Queen...   
6         M9B  Etobicoke  Cloverdale, Islington, Martin Grove, Princess ...   
7         M9C  Etobicoke  Bloordale Gardens, Eringate, Markland Wood, Ol...   
8         M9P  Etobicoke                                          Westmount   
9         M9R  Etobicoke  Kingsview Village, Martin Grove Gardens, Richv...   
10        M9V  Etobicoke  Albion Gardens, Beaumond Heights, Humbergate, ...   
11        M9W  Etobicoke                                          Northwest   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0   43.605647 -79.501321               3                   Gym   
1   43.602414 -79.543484               3           Pizza Place   
2   43.653654 -79.506944               1                  Pool   
3   43.636258 -79.498509               0        Baseball Field   
4   43.628841 -79.520999               3           Wings Joint   
6   43.650943 -79.554724               2                  Bank   
7   43.643515 -79.577201               3                  Café   
8   43.696319 -79.532242               3           Pizza Place   
9   43.688905 -79.554724               3     Mobile Phone Shop   
10  43.739416 -79.588437               3         Grocery Store   
11  43.706748 -79.594054               4             Drugstore   

   2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
0            Pizza Place                Bakery                       Café   
1                    Gym        Sandwich Place                Coffee Shop   
2                  River                  Park                Wings Joint   
3            Wings Joint    Chinese Restaurant        Fried Chicken Joint   
4          Grocery Store                Bakery               Burger Joint   
6            Wings Joint    Chinese Restaurant        Fried Chicken Joint   
7            Pizza Place     Convenience Store             Shopping Plaza   
8           Intersection        Sandwich Place  Middle Eastern Restaurant   
9                   Park              Bus Line                Pizza Place   
10   Fried Chicken Joint           Video Store             Sandwich Place   
11                   Bar   Rental Car Location                Wings Joint   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0   Fast Food Restaurant           Flower Shop   Fried Chicken Joint   
1               Pharmacy                   Pub                  Pool   
2   Fast Food Restaurant             Drugstore        Discount Store   
3            Flower Shop  Fast Food Restaurant             Drugstore   
4      Convenience Store        Discount Store  Fast Food Restaurant   
6            Flower Shop  Fast Food Restaurant             Drugstore   
7           Liquor Store            Beer Store                  Park   
8            Coffee Shop    Chinese Restaurant           Wings Joint   
9            Wings Joint    Chinese Restaurant  Fast Food Restaurant   
10          Liquor Store            Beer Store           Coffee Shop   
11                  Café           Flower Shop  Fast Food Restaurant   

   8th Most Common Venue 9th Most Common Venue  10th Most Common Venue  
0             Hobby Shop          Liquor Store                Pharmacy  
1           Skating Rink    Athletics & Sports                     Bar  
2      Convenience Store           Coffee Shop      Chinese Restaurant  
3         Discount Store     Convenience Store             Coffee Shop  
4                    Gym        Sandwich Place  Thrift / Vintage Store  
6         Discount Store     Conv

In [28]:
# create map
map_clusters = folium.Map(location=[TR_latitude, TR_longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(etobicoke_merged['Latitude'], etobicoke_merged['Longitude'], etobicoke_merged['Neighborhood'], etobicoke_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [29]:
etobicoke_merged.loc[etobicoke_merged['Cluster Labels'] == 0, etobicoke_merged.columns[[1] + list(range(5, etobicoke_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
3  Etobicoke               0        Baseball Field           Wings Joint   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
3    Chinese Restaurant   Fried Chicken Joint           Flower Shop   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
3  Fast Food Restaurant             Drugstore        Discount Store   

  9th Most Common Venue 10th Most Common Venue  
3     Convenience Store            Coffee Shop

In [30]:
etobicoke_merged.loc[etobicoke_merged['Cluster Labels'] == 1, etobicoke_merged.columns[[1] + list(range(5, etobicoke_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
2  Etobicoke               1                  Pool                 River   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
2                  Park           Wings Joint  Fast Food Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
2             Drugstore        Discount Store     Convenience Store   

  9th Most Common Venue 10th Most Common Venue  
2           Coffee Shop     Chinese Restaurant

In [31]:
etobicoke_merged.loc[etobicoke_merged['Cluster Labels'] == 2, etobicoke_merged.columns[[1] + list(range(5, etobicoke_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Etobicoke               2                  Bank           Wings Joint   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6    Chinese Restaurant   Fried Chicken Joint           Flower Shop   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6  Fast Food Restaurant             Drugstore        Discount Store   

  9th Most Common Venue 10th Most Common Venue  
6     Convenience Store            Coffee Shop

In [32]:
etobicoke_merged.loc[etobicoke_merged['Cluster Labels'] == 3, etobicoke_merged.columns[[1] + list(range(5, etobicoke_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0   Etobicoke               3                   Gym           Pizza Place   
1   Etobicoke               3           Pizza Place                   Gym   
4   Etobicoke               3           Wings Joint         Grocery Store   
7   Etobicoke               3                  Café           Pizza Place   
8   Etobicoke               3           Pizza Place          Intersection   
9   Etobicoke               3     Mobile Phone Shop                  Park   
10  Etobicoke               3         Grocery Store   Fried Chicken Joint   

   3rd Most Common Venue      4th Most Common Venue 5th Most Common Venue  \
0                 Bakery                       Café  Fast Food Restaurant   
1         Sandwich Place                Coffee Shop              Pharmacy   
4                 Bakery               Burger Joint     Convenience Store   
7      Convenience Store             Shopping Plaza          Liquor Store   
8         Sandwich Place  Middle Eastern Restaurant           Coffee Shop   
9               Bus Line                Pizza Place           Wings Joint   
10           Video Store             Sandwich Place          Liquor Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0            Flower Shop   Fried Chicken Joint            Hobby Shop   
1                    Pub                  Pool          Skating Rink   
4         Discount Store  Fast Food Restaurant                   Gym   
7             Beer Store                  Park              Pharmacy   
8     Chinese Restaurant           Wings Joint  Fast Food Restaurant   
9     Chinese Restaurant  Fast Food Restaurant             Drugstore   
10            Beer Store           Coffee Shop  Fast Food Restaurant   

   9th Most Common Venue  10th Most Common Venue  
0           Liquor Store                Pharmacy  
1     Athletics & Sports                     Bar  
4         Sandwich Place  Thrift / Vintage Store  
7              Drugstore          Discount Store  
8              Drugstore          Discount Store  
9         Discount Store       Convenience Store  
10           Pizza Place                Pharmacy

In [33]:
etobicoke_merged.loc[etobicoke_merged['Cluster Labels'] == 4, etobicoke_merged.columns[[1] + list(range(5, etobicoke_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke               4             Drugstore                   Bar   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11   Rental Car Location           Wings Joint                  Café   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11           Flower Shop  Fast Food Restaurant        Discount Store   

   9th Most Common Venue 10th Most Common Venue  
11     Convenience Store            Coffee Shop

## Well that was interesting
So it seems that clusters 0 and 2 are defined particularly by their one venue (the Baseball Field and the Bank). Clear trends aren't easily discernable except for Cluster 3, perhaps due to the fewness of locations provided from Foursquare for these neighborhoods in Etobicoke. Perhaps a different borough would be more interesting!